packages/react/src/components/Tree/Tree.tsx

```diff
@@ -0,0 +1,331 @@
+    return !_.isNil(item['items']) && item['items'].length > 0
+  }
+
+  static getItemsForRender = _.memoize((itemsFromProps: ShorthandCollection<TreeItemProps>) => {
+    const itemsForRenderGenerator = (
+      items = itemsFromProps,
+      level = 1,
+      parent?: ShorthandValue<TreeItemProps>,
+    ) => {
+      return _.reduce(
+        items,
+        (acc: Object, item: ShorthandValue<TreeItemProps>, index: number) => {
+          const id = item['id']
+          const isSubtree = Tree.hasSubtree(item)
+
+          acc[id] = {
+            elementRef: React.createRef<HTMLElement>(),
+            level,
+            index: index + 1, // Used for aria-posinset and it's 1-based.
+            parent,
+            siblings: items.filter(currentItem => currentItem !== item),
+          }
+
+          return {
+            ...acc,
+            ...(isSubtree ? itemsForRenderGenerator(item['items'], level + 1, item) : {}),
+          }
+        },
+        {},
+      )
+    }
+
+    return itemsForRenderGenerator(itemsFromProps)
+  })
+
+  static getAutoControlledStateFromProps(nextProps: TreeProps, prevState: TreeState) {
+    const itemsForRender = Tree.getItemsForRender(nextProps.items)
+
+    return {
+      itemsForRender,
+    }
+  }
+
+  getInitialAutoControlledState() {
+    return { activeItemIds: [] }
+  }
+
+  treeRef = React.createRef<HTMLElement>()
+
+  handleTreeItemOverrides = (predefinedProps: TreeItemProps) => ({
+    onTitleClick: (e: React.SyntheticEvent, treeItemProps: TreeItemProps) => {
+      let { activeItemIds } = this.state
+      const { id, siblings } = treeItemProps
+      const { exclusive } = this.props
+
+      if (!Tree.hasSubtree(treeItemProps)) {
+        return
+      }
+
+      const activeItemIdIndex = activeItemIds.indexOf(id)
+
+      if (activeItemIdIndex > -1) {
+        activeItemIds = [
+          ...activeItemIds.slice(0, activeItemIdIndex),
+          ...activeItemIds.slice(activeItemIdIndex + 1),
+        ]
+      } else {
+        if (exclusive) {
+          siblings.some(sibling => {
+            const activeSiblingIdIndex = activeItemIds.indexOf(sibling['id'])
+            if (activeSiblingIdIndex > -1) {
+              activeItemIds = [
+                ...activeItemIds.slice(0, activeSiblingIdIndex),
+                ...activeItemIds.slice(activeSiblingIdIndex + 1),
+              ]
+              return true
+            }
+            return false
+          })
+        }
+
+        activeItemIds = [...activeItemIds, id]
+      }
+
+      this.setState({
+        activeItemIds,
+      })
+
+      _.invoke(predefinedProps, 'onTitleClick', e, treeItemProps)
+    },
+    onFocusParent: (e: React.SyntheticEvent, treeItemProps: TreeItemProps) => {
+      const { parent } = treeItemProps
+
+      if (!parent) {
+        return
+      }
+
+      const { itemsForRender } = this.state
+      const elementToBeFocused = itemsForRender[parent['id']].elementRef
```

Comment:
would `itemsForRender[parent['id']]` never return NULL ? just to make sure it is safe